streams/src/main/java/org/apache/kafka/streams/state/internals/BoundsCheckingWindowStore.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.WindowStore;
+import org.apache.kafka.streams.state.WindowStoreIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BoundsCheckingWindowStore<K, V> implements WindowStore<K, V> {
```

Comment:
This is a meta comment: I think instead of adding this check on all the window store, we should only add this check for `SegmentedBytesStore`, since this is for this specific implementation which do not strictly obey window retention time. Imagine a user-customized window store implementation, then there is no need to do this extra check.

Reply:
Yeah, we can do this. I was thinking like this: "since the window retention time is part of the window definition, it should be enforce always, regardless of implementation." However, there is a nuance that I'm seeing now:
From a correctness standpoint, we only *need* to enforce the close time, which we're also doing in the appropriate Processor.
The other half is the duration for which it'll be queriable. I do think that it's fair game for us to enforce this universally, but I don't think it's that important, so we can implement it in the segment store instead.

Also worth noting: we were thinking strict enforcement was necessary for final updates, but it's actually not under the current design I'm working on, only exposing streamTime is.

I still think strict enforcement is nicer for developers than lax enforcement, though, so I'm still in favor of this PR.

Reply:
For final updates on the DSL, our enforcement on the processor should be sufficient, that no late records will be processed, and hence no emits after the strict deadline.

As you mentioned, enforcing this on the store as well is for 1) interactive queries and 2) PAPI users accessing the store directly. I think for those cases enforcing it universally might be an overkill. That being said, I do not have a concrete case in mind where a user customized window store would want to have lax enforcement, just felt it may be more flexible not enforcing, but I'm not feeling strong about it.

Reply:
I reflected on this, and I agree that what I've done here might be too agressive. I'll refactor and remove the `BoundsCheckingWindowStore`.

This will leave the enforcement in the processor and also will leave streamTime in the InternalProcessorContext. I don't think any additional logic in the SegmentStore would provide value.